File: src/support/typescript_language.cpp
#include "support/typescript_language.h"
#include "support/typescript.h"
#include <godot_cpp/core/error_macros.hpp>
#include <godot_cpp/core/memory.hpp>

const char *TypescriptLanguage::EXTENSION = "js";
const char *TypescriptLanguage::TYPE = "JavaScript";
TypescriptLanguage *TypescriptLanguage::singleton = nullptr;

TypescriptLanguage::TypescriptLanguage() {
	singleton = this;
}

TypescriptLanguage::~TypescriptLanguage() {
	singleton = nullptr;
	delete TYPE;
	delete EXTENSION;
}

TypescriptLanguage *TypescriptLanguage::get_singleton() {
	ERR_FAIL_NULL_V(singleton, nullptr);
	return singleton;
}

void TypescriptLanguage::_thread_enter() {
}

void TypescriptLanguage::_thread_exit() {
}

void TypescriptLanguage::_frame() {
}

String TypescriptLanguage::_get_name() const {
	return TYPE;
}

void TypescriptLanguage::_init() {
}

String TypescriptLanguage::_get_type() const {
	return TYPE;
}

String TypescriptLanguage::_get_extension() const {
	return EXTENSION;
}

void TypescriptLanguage::_finish() {
}

PackedStringArray TypescriptLanguage::_get_reserved_words() const {
	return PackedStringArray();
}

bool TypescriptLanguage::_is_control_flow_keyword(const String &keyword) const {
	return false;
}

PackedStringArray TypescriptLanguage::_get_comment_delimiters() const {
	return { "//" };
}

PackedStringArray TypescriptLanguage::_get_string_delimiters() const {
	return { "\" \"", "\' \'" };
}

TypedArray<Dictionary> TypescriptLanguage::_get_built_in_templates(const StringName &object) const {
	return TypedArray<Dictionary>();
}

bool TypescriptLanguage::_is_using_templates() {
	return false;
}

Dictionary TypescriptLanguage::_validate(const String &script, const String &path, bool validate_functions, bool validate_errors, bool validate_warnings, bool validate_safe_lines) const {
	Dictionary validation;
	validation["valid"] = true;
	return validation;
}

String TypescriptLanguage::_validate_path(const String &path) const {
	return "";
}

Object *TypescriptLanguage::_create_script() const {
	return memnew(Typescript);
}

bool TypescriptLanguage::_has_named_classes() const {
	return false;
}

bool TypescriptLanguage::_supports_builtin_mode() const {
	return false;
}

bool TypescriptLanguage::_supports_documentation() const {
	return false;
}

bool TypescriptLanguage::_can_inherit_from_file() const {
	return true;
}

int32_t TypescriptLanguage::_find_function(const String &class_name, const String &function_name) const {
	return 0;
}

String TypescriptLanguage::_make_function(const String &class_name, const String &function_name,
		const PackedStringArray &function_args) const {
	return "";
}

bool TypescriptLanguage::_overrides_external_editor() {
	return false;
}

Dictionary TypescriptLanguage::_complete_code(const String &code, const String &path, Object *owner) const {
	return Dictionary();
}

Dictionary TypescriptLanguage::_lookup_code(const String &code, const String &symbol, const String &path,
		Object *owner) const {
	return Dictionary();
}

String TypescriptLanguage::_auto_indent_code(const String &code, int32_t from_line, int32_t to_line) const {
	return "";
}

void TypescriptLanguage::_add_global_constant(const StringName &name, const Variant &value) {
}

void TypescriptLanguage::_add_named_global_constant(const StringName &name, const Variant &value) {
}

void TypescriptLanguage::_remove_named_global_constant(const StringName &name) {
}

Dictionary TypescriptLanguage::_debug_get_globals(int32_t max_subitems, int32_t max_depth) {
	return Dictionary();
}

void TypescriptLanguage::_reload_all_scripts() {
}

void TypescriptLanguage::_reload_tool_script(const Ref<Script> &script, bool soft_reload) {
}

PackedStringArray TypescriptLanguage::_get_recognized_extensions() const {
	PackedStringArray extensions;
	return { TypescriptLanguage::EXTENSION };
}

TypedArray<Dictionary> TypescriptLanguage::_get_public_functions() const {
	return TypedArray<Dictionary>();
}

Dictionary TypescriptLanguage::_get_public_constants() const {
	return Dictionary();
}

TypedArray<Dictionary> TypescriptLanguage::_get_public_annotations() const {
	return TypedArray<Dictionary>();
}

void TypescriptLanguage::_profiling_start() {
}

void TypescriptLanguage::_profiling_stop() {
}

bool TypescriptLanguage::_handles_global_class_type(const String &type) const {
	return false;
}

Dictionary TypescriptLanguage::_get_global_class_name(const String &path) const {
	return Dictionary();
}

Ref<Script> TypescriptLanguage::_make_template(const String &_template, const String &class_name,
		const String &base_class_name) const {
	Ref<Typescript> script;
	script.instantiate();
	Array values;
	values.append(class_name);
	values.append(base_class_name);
	String code = String(
			R"xxx(import {Node} from 'Node';

class {0} {

	_ready() {}

	_process(delta) {}

})xxx")
						  .format(values);
	script->set_source_code(code);
	return script;
}

File: src/support/typescript_language.js
import { ScriptLanguageExtension } from "src/js_godot/classes/script_language_extension";
import { PackedStringArray } from "src/js_godot/variant/packed_string_array";
import { GDString } from "src/js_godot/variant/gd_string";

export class TypeScriptLanguage extends ScriptLanguageExtension {
  constructor() {
    super();
  }

  _get_name() {
    return "TypeScript";
  }
  _init() {}
  _get_type() {
    return "TypeScript";
  }
  _get_extension() {
    return "ts";
  }
  _finish() {}
  _get_reserved_words() {}
  _is_control_flow_keyword(_keyword) {}
  _get_comment_delimiters() {}
  _get_doc_comment_delimiters() {}
  _get_string_delimiters() {
    return new PackedStringArray([new GDString('""'), new GDString("''")]);
  }
  _make_template(_template, _class_name, _base_class_name) {}
  _get_built_in_templates(_object) {}
  _is_using_templates() {}
  _validate(
    _script,
    _path,
    _validate_functions,
    _validate_errors,
    _validate_warnings,
    _validate_safe_lines
  ) {}
  _validate_path(_path) {}
  _create_script() {}
  _has_named_classes() {}
  _supports_builtin_mode() {}
  _supports_documentation() {}
  _can_inherit_from_file() {}
  _find_function(_function, _code) {}
  _make_function(_class_name, _function_name, _function_args) {}
  _can_make_function() {}
  _open_in_external_editor(_script, _line, _column) {}
  _overrides_external_editor() {}
  _preferred_file_name_casing() {}
  _complete_code(_code, _path, _owner) {}
  _lookup_code(_code, _symbol, _path, _owner) {}
  _auto_indent_code(_code, _from_line, _to_line) {}
  _add_global_constant(_name, _value) {}
  _add_named_global_constant(_name, _value) {}
  _remove_named_global_constant(_name) {}
  _thread_enter() {}
  _thread_exit() {}
  _debug_get_error() {}
  _debug_get_stack_level_count() {}
  _debug_get_stack_level_line(_level) {}
  _debug_get_stack_level_function(_level) {}
  _debug_get_stack_level_source(_level) {}
  _debug_get_stack_level_locals(_level, _max_subitems, _max_depth) {}
  _debug_get_stack_level_members(_level, _max_subitems, _max_depth) {}
  _debug_get_stack_level_instance(_level) {}
  _debug_get_globals(_max_subitems, _max_depth) {}
  _debug_parse_stack_level_expression(
    _level,
    _expression,
    _max_subitems,
    _max_depth
  ) {}
  _debug_get_current_stack_info() {}
  _reload_all_scripts() {}
  _reload_tool_script(_script, _soft_reload) {}
  _get_recognized_extensions() {
    return new PackedStringArray([new GDString("ts"), new GDString("js")]);
  }
  _get_public_functions() {}
  _get_public_constants() {}
  _get_public_annotations() {}
  _profiling_start() {}
  _profiling_stop() {}
  _profiling_set_save_native_calls(_enable) {}
  _profiling_get_accumulated_data(_info_array, _info_max) {}
  _profiling_get_frame_data(_info_array, _info_max) {}
  _frame() {}
  _handles_global_class_type(_type) {}
  _get_global_class_name(_path) {}
}
